### PR TITLE
Rebased iterator docs

### PR DESCRIPTION
**Pull request by PragTob**
This is based on and should deprecate #1063 :)

I took the original PR, rebased it against current master and rewrote the documentation so that it applies to iterator. It only includes the original docs right now, I'm keen on adding more either in this PR or maybe discuss and merge here and then add more docs in a separate PR.

Something that has been a bit difficult to me here is clearly formulating the method documentation. The iterator methods all return a new iterator instance that then does what? For instance reject, I went with:

> The returned iterator only returns elements for which the the block returned a falsey value.

Does that seem good? Do we want something different?

cc: @davydovanton

### CODE REVIEW

**Comment by jhass** [comment]
Yes, currently I see three styles here:

1. Does X
2. Returns an iterator that does X
3. The returned iterator does X

It would be nice to pick one and go with it, I have no clear favourite but would lean towards 2.

File: src/iterator.cr [hunk 1]
```diff
@@ -639,6 +727,17 @@ module Iterator(T)
     end
   end
 
+  # Returns an iterator that takes puts together the corresponding elements
```
Comment (jhass) [inline src/iterator.cr]: So for the style 2 version I'd go with "Returns an iterator that returns the elements of this iterator and the passed one pairwise."

**Comment by PragTob** [comment]
Seems good to me, but I also problems saying what "X" is. Is it "returns value that do ..." or "yields values that ..." - yield is confusing as it returns them when you call next, but I don't want to get the `next` thing in there as that is the basic how iterators work.

**Comment by jhass** [comment]
"return" seems more clear indeed.

File: src/iterator.cr [hunk 2]
```diff
@@ -258,6 +270,17 @@ module Iterator(T)
...
+  # back at the beginning when the end was reached, but only n times.
+  #
+  #     iter = ["a", "b", "c"].each.cycle(2)
+  #     iter.next # => "a"
+  #     iter.next # => "b"
+  #     iter.next # => "c"
+  #     iter.next # => "a"
+  #     iter.next # => "b"
+  #     iter.next # => "c"
+  #     iter.next # => Iterator::Stop
```
Comment (PragTob) [inline src/iterator.cr]: This should also be `Iterator::Stop::INSTANCE` - doh

**Comment by davydovanton** [comment]
I liked this. @PragTob thanks for initiative :+1:

**Comment by PragTob** [comment]
Updated trying to unify it and cleanup my `INSTANCE` mistake :)

**Comment by jhass** [comment]
I like it, thank you :+1: 

Mind squashing your two commits together? `git rebase -i HEAD^^`, pick squash for the second entry, `git push -f` to your branch.

**Comment by PragTob** [comment]
Nope don't mind squashing them together, gotta run now will do it later thanks :)

**Comment by jhass** [comment]
Merged your two commits as 13f838ac1afc4a577459633c66dea86045d3092b

Thanks!

**Comment by davydovanton** [comment]
:tada:

**Comment by PragTob** [comment]
Thanks @jhass (was just going to do that :D )

**Pull request closed without merging**